### PR TITLE
[Storage] Fix #20638: `az storage account\container\blob generate-sas`: Migrate to track2 sdk to allow new permissions

### LINKED ISSUE #20368: SAS token for Blob Index Tags
**Is your feature request related to a problem? Please describe.**

Azure CLI doesn't support [Blob Index Tags](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-manage-find-blobs?tabs=azure-portal#account-sas) permissions `t f x`. It throws validation error.

```azcli
az storage account generate-sas --account-key xxx --account-name myaccount --expiry 2020-01-01 --https-only --permissions rwdlactfx --resource-types cos --services b

valid values are (a)dd (c)reate (d)elete (l)ist (p)rocess (r)ead (u)pdate (w)rite (a)dd (c)reate (d)elete (l)ist (p)rocess (r)ead (u)pdate (w)rite or a combination thereof
```

**Describe the solution you'd like**

Add `t` `f` and `x` to acceptable values for `--permissions`.

**Describe alternatives you've considered**

**Additional context**

Works with `PowerShell` and is [well documented with `REST`](https://docs.microsoft.com/en-us/rest/api/storageservices/create-account-sas?redirectedfrom=MSDN#specifying-account-sas-parameters).
```PowerShell
$storage = Get-AzStorageAccount -Name myaccount -ResourceGroupName mygroup
New-AzStorageAccountSASToken -Service Blob -ResourceType Service,Container,Object -Permission rwdlactfx -Protocol HttpsOnly -StartTime (Get-Date) -ExpiryTime (Get-Date).AddDays(1) -Context $storage.context
```

### PR DESCRIPTION
**Pull request by calvinhzy**
**Description**

Fix #20368
Migrate to track2 sdk for `az storage account\container\blob generate-sas`
This allows new permissions to be added.
Account permissions: (a)dd (c)reate (d)elete (f)ilter_by_tags (i)set_immutability_policy (l)ist (p)rocess (r)ead (t)ag (u)pdate (w)rite (x)delete_previous_version (y)permanent_delete.
Container permissions: (a)dd (c)reate (d)elete (e)xecute (f)ilter_by_tags (i)set_immutability_policy (l)ist (m)ove (r)ead (t)ag (w)rite (x)delete_previous_version (y)permanent_delete.
Blob permissions: (a)dd (c)reate (d)elete (e)xecute (i)set_immutability_policy (m)ove (r)ead (t)ag (w)rite (x)delete_previous_version (y)permanent_delete.

**Testing Guide**

`az storage account generate-sas --account-key xxx --account-name myaccount --expiry 2020-01-01 --https-only --permissions rwdlactfx --resource-types cos --services b`
It should show the account sas token
`az storage blob exists -c {container} -n {blob} --account-name {account_name} --sas-token {account_sas}` 

`az storage container generate-sas -n {container} --https-only --permissions racwdxyltfmei --connection-string {con_str} --expiry {expiry}`
`az storage blob upload -c {container} -f "{local_file}" -n {blob} --account-name {account} --sas-token "{container_sas}"`

`az storage blob generate-sas -c {container} -n {blob} --account-name {account} --https-only --permissions racwdxytmei --expiry {expiry}`
`az storage blob show -c {container} -n {blob} --account-name {account} --sas-token {blob_sas}`

**History Notes**

[Storage] `az storage account\container\blob generate-sas`: Allow new permissions

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Storage

File: src/azure-cli/azure/cli/command_modules/storage/operations/blob.py [hunk 1]
```diff
@@ -728,47 +728,88 @@ def _delete_blob(blob_name):
...
-        sas_token = client.generate_blob_shared_access_signature(
-            container_name, blob_name, permission=permission, expiry=expiry, start=start, id=id, ip=ip,
-            protocol=protocol, cache_control=cache_control, content_disposition=content_disposition,
-            content_encoding=content_encoding, content_language=content_language, content_type=content_type)
+        account_key = client.credential.account_key
+
+    blob_url = kwargs.pop('blob_url')
+    container_name = kwargs.pop('container_name')
+    blob_name = kwargs.pop('blob_name')
+    if blob_url:
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/storage/operations/blob.py:763]: I think we don't need to support `--blob-url` for `generate-sas` command. It's find to keep arguments same with track2. Adding `--blob-url` support makes things too complicated.

File: src/azure-cli/azure/cli/command_modules/storage/_params.py [hunk 2]
```diff
@@ -796,11 +798,12 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
...
                    help='Show nextMarker in result when specified.')
 
-    with self.argument_context('storage blob generate-sas') as c:
+    with self.argument_context('storage blob generate-sas', resource_type=ResourceType.DATA_STORAGE_BLOB) as c:
         from .completers import get_storage_acl_name_completion_list
 
-        t_blob_permissions = self.get_sdk('blob.models#BlobPermissions')
+        t_blob_permissions = self.get_sdk('_models#BlobSasPermissions', resource_type=ResourceType.DATA_STORAGE_BLOB)
         c.register_sas_arguments()
+        c.register_blob_arguments_track2()
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/storage/_params.py:806]: So we can just use `c.register_blob_arguments()` here

File: src/azure-cli/azure/cli/command_modules/storage/operations/blob.py [hunk 3]
```diff
@@ -728,47 +728,88 @@ def _delete_blob(blob_name):
...
+                                    snapshot=snapshot, account_key=account_key, user_delegation_key=user_delegation_key,
+                                    permission=permission, expiry=expiry, start=start, policy_id=id, ip=ip,
+                                    protocol=protocol, cache_control=cache_control,
+                                    content_disposition=content_disposition, content_encoding=content_encoding,
+                                    content_language=content_language, content_type=content_type, **kwargs)
+
     if full_uri:
-        return encode_url_path(client.make_blob_url(container_name, blob_name, protocol=protocol,
-                                                    sas_token=quote(sas_token, safe='&%()$=\',~')))
+        return encode_url_path(make_blob_url(url=blob_url, protocol=protocol,
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/storage/operations/blob.py:784]: Why not use `blob_client.url` but manually implemented `make_blob_url`?

Comment (calvinhzy) [reply src/azure-cli/azure/cli/command_modules/storage/operations/blob.py:784]: It combines the sas_token with the url.

**Review by evelyn-ys — APPROVED** [review]

**Pull request merged**